includes/class-llms-elementor-migrate.php

```diff
@@ -0,0 +1,273 @@
+				),
+			),
+			'isInner'  => false,
+		);
+		$content[] = array(
+			'id'       => uniqid(),
+			'elType'   => 'container',
+			'settings' => array(),
+			'elements' => array(
+				array(
+					'id'         => uniqid(),
+					'elType'     => 'widget',
+					'settings'   => array(),
+					'elements'   => array(),
+					'widgetType' => 'llms_course_instructors_widget',
+				),
+			),
+			'isInner'  => false,
+		);
+		$content[] = array(
+			'id'       => uniqid(),
+			'elType'   => 'container',
+			'settings' => array(),
+			'elements' => array(
+				array(
+					'id'         => uniqid(),
+					'elType'     => 'widget',
+					'settings'   => array(),
+					'elements'   => array(),
+					'widgetType' => 'llms_pricing_table_widget',
+				),
+			),
+			'isInner'  => false,
+		);
+		$content[] = array(
+			'id'       => uniqid(),
+			'elType'   => 'container',
+			'settings' => array(),
+			'elements' => array(
+				array(
+					'id'         => uniqid(),
+					'elType'     => 'widget',
+					'settings'   => array(),
+					'elements'   => array(),
+					'widgetType' => 'llms_course_progress_widget',
+				),
+			),
+			'isInner'  => false,
+		);
+		$content[] = array(
+			'id'       => uniqid(),
+			'elType'   => 'container',
+			'settings' => array(),
+			'elements' => array(
+				array(
+					'id'         => uniqid(),
+					'elType'     => 'widget',
+					'settings'   => array(),
+					'elements'   => array(),
+					'widgetType' => 'llms_course_continue_button_widget',
+				),
+			),
+			'isInner'  => false,
+		);
+		$content[] = array(
+			'id'       => uniqid(),
+			'elType'   => 'container',
+			'settings' => array(),
+			'elements' => array(
+				array(
+					'id'         => uniqid(),
+					'elType'     => 'widget',
+					'settings'   => array(),
+					'elements'   => array(),
+					'widgetType' => 'llms_course_syllabus_widget',
+				),
+			),
+			'isInner'  => false,
+		);
+
+		return $content;
+	}
+
+	/**
+	 * Migrate posts created prior to the elementor updates to have default LifterLMS widgets.
+	 *
+	 * @since [version]
+	 *
+	 * @return  void
+	 */
+	public function migrate_post() {
+
+		global $pagenow;
+
+		if ( 'post.php' !== $pagenow ) {
+			return;
+		}
+
+		$post_id = llms_filter_input( INPUT_GET, 'post', FILTER_SANITIZE_NUMBER_INT );
+		$post    = $post_id ? get_post( $post_id ) : false;
+
+		if ( ! $post || ! isset( $_REQUEST['action'] ) || 'elementor' !== $_REQUEST['action'] || ! $this->should_migrate_post( $post_id ) || 'course' !== get_post_type( $post_id ) ) {
+			return;
+		}
+
+		$this->ensure_elementor_data_present( $post_id );
+		$this->add_template_to_post( $post_id );
+	}
+
+	public function add_template_to_post( $post_id ) {
+		$content = get_post_meta( $post_id, '_elementor_data', true );
+		if ( ! $content ) {
+			return;
+		}
+
+		$content = json_decode( $content, true );
+
+		if ( ! is_array( $content ) ) {
+			return;
+		}
+
+		$content = array_merge( $content, $this->get_elementor_data_template() );
```

Comment:
As far as I can tell we throw away the default (core WP) post_content.

There were some reasons for doing this. I believe one is that for the LifterLMS blocks in particular, you might get content like this, which includes the WP block editor HTML comment and some preview HTML that is kind of frozen in time and not kept up to date.

````
<!-- wp:llms/course-continue-button -->
<div class="wp-block-llms-course-continue-button" style="text-align: center;">[lifterlms_course_continue_button]</div>
<!-- /wp:llms/course-continue-button -->
````

I'm worried about 2 things in particular.

1. We also remove any content the user added to the course description (the_content) in addition to what LifterLMS puts in there, including some blocks/etc (like paragraph, headings, shortcode, video, and image blocks) that do convert decently to Elementor this way.

2. Users of Elementor will expect some amount of the old content to transfer over when they click Edit with Elementor. It works this way for some other course CPTs (PMPro Courses e.g.) and I feel users will see this as a deficiency in LifterLMS vs the subtle issue it is. And ultimately, users who have already edited course pages will have to go back and forth between the WP and Elementor versions of the post to copy/paste things, if they even know that they can do that.

So I think we need to get "clever" here and try our best to (a) preserve the_content that is there while also (b) removing or addressing the bits that are going to be confusing because they don't work as intended.

A simple solution would be to grab the_content and use REGEX or HTML parsing to remove anything between HTML comments like `<!-- wp:llms/.*<!-- /wp:llms/.*-->` (need the actual REGEX)

Thoughts?

Reply:
I believe until you save the post in Elementor, the_content is preserved. I’d be curious to know what block content from PMPro CPTs is preserved when starting to edit with elementor.

We could look into adding a widget with the content from the_content as a widget at the top if there is no content in `_elementor_data`, which I believe is all that Elementor does when you first edit.

Reply:
I've changed it to use `post_content` as the default for `_elementor_data` rather than an empty array. It also uncovered a bug where it needed to use `wp_slash` before the call to `update_post_meta` to avoid the json becoming unparsable with text editor content such as this.